### PR TITLE
Rosbridge websocket client mode

### PR DESCRIPTION
**Pull request by jihoonl**
Opposite to rosbridge_server, it connects to existing webserver to expose its ros api.

### CODE REVIEW

**Comment by viktorku** [comment]
Can you shortly explain why and when would we need this? Is it to propagate ROS constructs created with roslibjs to the server? Like if we advertise and publish on a topic in the browser a backend node to pick those messages up?

**Comment by jihoonl** [comment]
Often, I had use cases that require to control robots behind firewall(robots that have internet access but don't have public IP address) via web portal in cloud. Something like the drawing below.

untitled presentation

To initiate the channel between webserver and robots behind firewall, rosbridge_server wasn't an option because webserver cannot open a websocket channel unless robot's url is static and publically available.

`rosbridge_client` is introduced to fulfill this usecase. Opposite to the rosbridge_server which just waits for connection, it initiates the connection to the webserver then receives rosbridge_protocol messages.

File: rosbridge_client/src/rosbridge_client/websocket_handler.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+        rospy.loginfo("connected.")
+
+    def closed(self, code, reason):
+        """
+        Called by the server when the websocket stream and connection are
+        finally closed
+        """
+        # rospy.loginfo("Closed")
+        self.protocol.finish()
+        print("Closed. (%s, %s)" %( code, reason))
```
Comment (Behery) [inline rosbridge_client/src/rosbridge_client/websocket_handler.py]: Why not rospy.loginfo?
```suggestion
        def closed(self, code, reason):
            """
            Called by the server when the websocket stream and connection are
            finally closed
            """
            self.protocol.finish()
            rospy.loginfo("Closed. (%s, %s)" %( code, reason))
```

File: rosbridge_client/src/rosbridge_client/websocket_handler.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+            "delay_between_messages": cls.delay_between_messages,
+            "max_message_size": cls.max_message_size,
+            "unregister_timeout": cls.unregister_timeout,
+            "bson_only_mode": cls.bson_only_mode
+        }
+        try:
+            self.protocol = RosbridgeProtocol(cls.seed, parameters=parameters)
+            self.protocol.outgoing = self.send_message
+            self.authenticated = False
+        except Exception as exc:
```
Comment (Behery) [inline rosbridge_client/src/rosbridge_client/websocket_handler.py]: `pass` is not needed here

```suggestion
            except Exception as exc:
                rospy.logerr("Unable to accept incoming connection.  Reason: %s",
                         str(exc))
```

File: rosbridge_client/test/simple_test_server.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+        msg = {
+            "op": "advertise",
+            "topic": "/take_off",
+            "type": "std_msgs/String"
+        }
+        self.send_message(json.dumps(msg))
+        print("Advertise %s" % msg)
+
+    def _publish(self):
+        msg = { "op": "publish",
```
Comment (viktorku) [inline rosbridge_client/test/simple_test_server.py]: Fix indentation by moving to next line

File: rosbridge_client/test/simple_test_server.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+                "topic": "/take_off",
+                "msg": {
+                    "data": "Take off!!"
+                }
+            }
+        self.send_message(json.dumps(msg))
+        print("Publish %s" % msg)
+
+    def _subscribe(self):
+        msg = { "op": "subscribe",
```
Comment (viktorku) [inline rosbridge_client/test/simple_test_server.py]: Same here as above

File: rosbridge_client/test/simple_test_server.py [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+    app = Application([(r"/", Handler), (r"", Handler)])
+    connected = False
+    address = 'localhost'
+    port = 9090
+
+    num_try = 0
+    while not connected and num_try < 5:
+        try:
+            app.listen(port, address)
+            print("Started on port %d" % port)
```
Comment (viktorku) [inline rosbridge_client/test/simple_test_server.py]: print statement can be replaced with `rospy.loginfo`

File: rosbridge_client/test/simple_test_server.py [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+            connected = True
+        except error as e:
+            print("Failed to start server retry...")
```
Comment (viktorku) [inline rosbridge_client/test/simple_test_server.py]: `rospy.logerr`

File: rosbridge_client/README.md [hunk 7]
```diff
@@ -0,0 +1,3 @@
+# Rosbridge Client
+
+It provides client mode of rosbridge which connects to the server as a client.
```
Comment (viktorku) [inline rosbridge_client/README.md]: Perhaps most of the contents of your comment https://github.com/RobotWebTools/rosbridge_suite/pull/360#issuecomment-431278180 can be moved in this README. The graph image helps being included here as well.

**Comment by jihoonl** [comment]
https://github.com/ros/rosdistro/pull/19352

**Comment by viktorku** [comment]
> ros/rosdistro#19352 to fix travis

`python-ws4py` should be included as a dependee in `package.xml` as well.

File: rosbridge_client/README.md (hunk 7 above)
Comment (nageshpindi) [reply rosbridge_client/README.md]: which ROSBRDIGE version are u using.....?

**Comment by karthikdesai** [comment]
*bump* Any updates on this? :)

**Comment by jihoonl** [comment]
I will reopen when I have a time to follow up til merge.

**Comment by JuuuoPeter** [comment]
I can use client to connect the server,however i can't communicate with each other by topics,servers...

**Pull request closed without merging**